src/devices/Display/LedDriver/Ht16k33.cs

```diff
@@ -0,0 +1,212 @@
+        /// Initialize HT16K33 device
+        /// </summary>
+        /// <param name="i2cDevice">The <see cref="System.Device.I2c.I2cDevice"/> to create with.</param>
+        public Ht16k33(I2cDevice i2cDevice)
+        {
+            // Initialize I2C device
+            I2cDevice = i2cDevice;
+
+            TurnOnOscillator();
+            SetBrightness();
+            SetBlinkDisplayOn();
+        }
+
+        /// <summary>
+        /// Gets or sets level of display brightness
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Brightness must be between 0 and <see cref="MaxBrightness"/></exception>
+        public byte Brightness
+        {
+            get => _brightness;
+            set
+            {
+                if (value == _brightness)
+                {
+                    return;
+                }
+
+                if (value > MaxBrightness)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(Brightness), $"Brightness must be between 0 and {MaxBrightness}");
+                }
+
+                _brightness = value;
+
+                SetBrightness();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets display blink rate, also turns on display if turned off
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException" />
+        public BlinkRate BlinkRate
+        {
+            get => _blinkRate;
+            set
+            {
+                if (value == _blinkRate)
+                {
+                    return;
+                }
+
+                if (!Enum.IsDefined(typeof(BlinkRate), _blinkRate))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(_blinkRate));
+                }
+
+                _blinkRate = value;
+
+                SetBlinkDisplayOn();
+            }
+        }
+
+        /// <summary>
+        /// Sets screen on/off
+        /// </summary>
+        public bool DisplayOn
+        {
+            get => _displayOn;
+            set
+            {
+                if (_displayOn = value)
+                {
+                    SetBlinkDisplayOn();
+                }
+                else
+                {
+                    I2cDevice.WriteByte((byte)Command.Blink);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            I2cDevice?.Dispose();
+            I2cDevice = null;
+        }
+
+        #region Abstract members
+        /// <summary>
+        /// Write raw data to display buffer
+        /// </summary>
+        /// <param name="data">Array of bytes to write to the display</param>
+        /// <param name="startAddress">Address to start writing from</param>
+        public abstract void Write(ReadOnlySpan<byte> data, int startAddress = 0);
+
+        /// <summary>
+        /// Clear display buffer
+        /// </summary>
+        public abstract void Clear();
```

Comment:
why are these abstract, Htk16k33 seems pretty specific

Reply:
As I mentioned before this chip is used in multiple configurations of which I have only one hardware. After examining the available code samples for other hardware using this chipset I've come to the conclusion that the common features consist of the setup commands (turn on oscillator, set brightness, blinking), but the actual display memory layout can vary significantly. It could be a segmented display or an LED matrix that requires different coding altogether imho.